PowerEditor/src/NppBigSwitch.cpp

```diff
@@ -2089,22 +2103,147 @@ LRESULT Notepad_plus::process(HWND hwnd, UINT message, WPARAM wParam, LPARAM lPa
 						wmqesType += " - ENDSESSION_LOGOFF";
 				}
 				string queryEndSession = "WM_QUERYENDSESSION (lParam: " + wmqesType + ") =====================================";
-				if (WM_QUERYENDSESSION == message)
-					writeLog(nppIssueLog.c_str(), queryEndSession.c_str());
+				writeLog(nppIssueLog.c_str(), queryEndSession.c_str());
+			}
+
+			// here is also the right place to unblock a possible modal-dlg blocking the main N++ wnd
+			// (it is then too late to do so at the WM_ENDSESSION time, as for this we need this thread message queue...)
+			if (!::IsWindowEnabled(hwnd))
+			{
+				// we probably have a blocking modal-window like MessageBox (e.g. the "Reload" or "Keep non existing file")
+				if (!nppParam.isEndSessionCritical() && isFirstQueryEndSession)
+					return FALSE; // request abort of the shutdown (for a non-critical one we can give the user a chance to solve whatever is needed)
+
+				// in most cases we will need to take care and programmatically close such dialogs in order to exit gracefully,
+				// otherwise the N++ most probably crashes itself without any tidy-up
+
+				string strLog = "Main N++ wnd is disabled by (an active modal-dlg?):  ";
+				char szBuf[MAX_PATH + 128] = { 0 };
+
+				HWND hActiveWnd = ::GetActiveWindow();
+				if (hActiveWnd)
+				{
+					if (::GetWindowTextA(hActiveWnd, szBuf, _countof(szBuf)))
+						strLog += szBuf;
+					::SendMessage(hActiveWnd, WM_CLOSE, 0, 0);
+				}
 				else
-					writeLog(nppIssueLog.c_str(), "WM_CLOSE (isQueryEndSessionStarted == true)");
+				{
+					// no active child window, so it is most probably the system dialog box class #32770 (used e.g. by the MessageBox WINAPI)
+					// - so our main hwnd here is not the PARENT but OWNER of that top-level window
+					hActiveWnd = ::GetLastActivePopup(hwnd);
+					if (hActiveWnd)
+					{
+						if (::GetWindowTextA(hActiveWnd, szBuf, _countof(szBuf)))
+							strLog += szBuf;
+						::GetClassNameA(hActiveWnd, szBuf, _countof(szBuf));
+						if (0 == lstrcmpiA("#32770", szBuf))

```

Comment:
Please respect the coding convention:
use `if (lstrcmpiA("#32770", szBuf) == 0)` instead.

Reply:
Sorry for that, force of habit, I will change all these to the N++ style.

Once I had a bad time with a `if (variable=VALUE)` typo which was supposed to be `if (variable==VALUE)`. Since then I have learned the opposite way because of the compilers do not allow `if (VALUE=variable)`...